torch/_dynamo/variables/misc.py

```diff
@@ -234,11 +234,44 @@ def visit(node):
 
         VariableTracker.apply(visit, (args, kwargs))
 
+        ctx = AutogradFunctionContextVariable.create(tx)
+        args = [ctx, *args]
+
         if requires_grad and torch.is_grad_enabled():
-            # TODO(jansel): handle this in training mode
-            unimplemented("autograd.Function with requires_grad")
+            from .torch import is_fn_safe_to_run, TorchHigherOrderOperator
+
+            def trampoline_autograd_apply(*args, **kwargs):
+                return self.fn_cls.apply(*args, **kwargs)
+
+            def trampoline_autograd_fwd(*args, **kwargs):
+                return self.fn_cls.forward(*args, **kwargs)
+
+            def trampoline_autograd_bwd(*args, **kwargs):
+                return self.fn_cls.backward(*args, **kwargs)
+
+            # Speculate fwd, will raise unimplemented and bubble up if not sound, or will accumulate guards
+            # onto tx if sound.
+            # TODO(voz): NOTE: This function identity is unguarded, but the odds of someone swapping self.fn_cls from
+            # autograd fn to something else is very low.
+            higher_order_autograd_fn = TorchHigherOrderOperator(trampoline_autograd_fwd)
+            speculated_fwd_result = higher_order_autograd_fn.call_function(
+                tx, args, kwargs
+            )
+            bwd_args = [ctx, speculated_fwd_result]
+            # ctx.value.saved_tensors = ctx.value.to_save
+            if not is_fn_safe_to_run(
+                tx, TorchHigherOrderOperator(trampoline_autograd_bwd), bwd_args
+            ):
+                unimplemented("Unsafe bwd in autograd.function")
+
+            # If fwd and backward are sound, we want apply in the graph.
```

Comment:
Not sure about this comment: we want apply in the graph because we need to preserve the autograd behavior for AOTAutograd to be able to do its job.

Reply:
Right, I think we are saying the same thing! We *must* have apply in the graph, but we need to check the fwd and bwd to ensure it is sound.

Reply:
Yes we agree on what needs to be done (and is done).

> because doing so messes with the versioning of tensors for grad in bwd.

That is the part that is confusing to me as a justification.

Reply:
Ah, agreed - if we just write forward in the graph, you don't get all of apply, and therefore, no backward. This is just me musing as I learn how it works, its a low value comment.